{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "private_outputs": true,
      "provenance": [],
      "mount_file_id": "1TgC0Cd6R7STlydfr1Vg95vDVFpiUBnFI",
      "authorship_tag": "ABX9TyN9FzOKZBoQM3v5465XxnpK",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Palaeoprot/ProteoParc_Colab/blob/main/PalaeoParc_in_Colab.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##License\n",
        "\n",
        "The [ProteoParc](https://github.com/guillecarrillo/proteoparc) pipeline was written by [Guillermo Carrillo Mart√≠n](https://github.com/guillecarrillo) and this is his license from Github. \"[Palaeoparc] \"*is the result of many collective work, so its citation in any study in which it is used would be appreciated.\n",
        "\n",
        "\n",
        " Every user is free to change any section of the code so it can be adjusted to individual necessities, if you do so, please indicate it in your manuscripts writing the changes to avoid confusion. For more information, read the License This pi[link text](https://)peline has been possible thanks to a lot of collective work, so its citation in any study in which it is used would be appreciated. Every user is free to change any section of the code so it can be adjusted to individual necessities, if you do so, please indicate it in your manuscripts to avoid confusion.*\"\n",
        "\n",
        "Here is my take on it in Colab\n",
        "\n",
        "<img src='https://drive.google.com/uc?export=view&id=1FqSr3E9UQKtRO4KwR6LjAU8TPrD8Oq6q' width=800px align=centre>\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "73G4TMMXt0TO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython\n",
        "import requests, sys, json, os, time  #data handling\n",
        "import argparse  # command-line arguments\n",
        "import shlex  # parsing shell commands\n",
        "import subprocess  # For system commands (already imported twice)\n",
        "import ipywidgets as widgets  # For interactive elements\n",
        "from IPython.display import display, HTML, clear_output  # For interactive output\n",
        "import shutil\n",
        "import re\n",
        "import os\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from requests.adapters import HTTPAdapter, Retry\n",
        "from datetime import datetime\n",
        "\n",
        "# Mount Google Drive (Optional, you can comment out if not using)\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "# Global constants\n",
        "WEBSITE_API = \"https://rest.uniprot.org/\""
      ],
      "metadata": {
        "id": "xJAaQwqisbMX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "bone_proteins = {\n",
        "    \"ALB\": \"serum albumin\",\n",
        "    \"AEBP1\": \"AE binding protein 1\",\n",
        "    \"AHSG\": \"alpha-2-HS-glycoprotein (fetuin-A)\",\n",
        "    \"ALPL\": \"alkaline phosphatase, liver/bone/kidney\",\n",
        "    \"APOA1\": \"apolipoprotein A-I\",\n",
        "    \"APOA4\": \"apolipoprotein A-IV\",\n",
        "    \"APOE\": \"apolipoprotein E\",\n",
        "    \"APOC1\": \"apolipoprotein C-I\",\n",
        "    \"APP\": \"amyloid beta precursor protein\",\n",
        "    \"ASPN\": \"asporin\",\n",
        "    \"BGLAP\": \"bone gamma-carboxyglutamate protein (osteocalcin)\",\n",
        "    \"BGN\": \"biglycan\",\n",
        "    \"C3\": \"complement component 3\",\n",
        "    \"C8B\": \"complement component 8, beta polypeptide\",\n",
        "    \"C9\": \"complement component 9\",\n",
        "    \"CFH\": \"complement factor H\",\n",
        "    \"CHAD\": \"chondroadherin\",\n",
        "    \"CHGA\": \"chromogranin A\",\n",
        "    \"CLEC11A\": \"C-type lectin domain family 11, member A\",\n",
        "    \"CLEC3B\": \"C-type lectin domain family 3, member B\",\n",
        "    \"CLU\": \"clusterin\",\n",
        "    \"COL10A1\": \"collagen type X alpha 1 chain\",\n",
        "    \"COL11A1\": \"collagen type XI alpha 1 chain\",\n",
        "    \"COL11A2\": \"collagen type XI alpha 2 chain\",\n",
        "    \"COL12A1\": \"collagen type XII alpha 1 chain\",\n",
        "    \"COL16A1\": \"collagen type XVI alpha 1 chain\",\n",
        "    \"COL1A1\": \"collagen type I alpha 1 chain\",\n",
        "    \"COL1A2\": \"collagen type I alpha 2 chain\",\n",
        "    \"COL21A1\": \"collagen type XXI alpha 1 chain\",\n",
        "    \"COL22A1\": \"collagen type XXII alpha 1 chain\",\n",
        "    \"COL2A1\": \"collagen type II alpha 1 chain\",\n",
        "    \"COL3A1\": \"collagen type III alpha 1 chain\",\n",
        "    \"COL4A3\": \"collagen type IV alpha 3 chain\",\n",
        "    \"COL4A4\": \"collagen type IV alpha 4 chain\",\n",
        "    \"COL4A5\": \"collagen type IV alpha 5 chain\",\n",
        "    \"COL5A1\": \"collagen type V alpha 1 chain\",\n",
        "    \"COL5A2\": \"collagen type V alpha 2 chain\",\n",
        "    \"COL5A3\": \"collagen type V alpha 3 chain\",\n",
        "    \"COL6A1\": \"collagen type VI alpha 1 chain\",\n",
        "    \"COL6A3\": \"collagen type VI alpha 3 chain\",\n",
        "    \"COL8A1\": \"collagen type VIII alpha 1 chain\",\n",
        "    \"CRP\": \"C-reactive protein\",\n",
        "    \"DCN\": \"decorin\",\n",
        "    \"DPT\": \"dermatopontin\",\n",
        "    \"EEF1A1\": \"eukaryotic translation elongation factor 1 alpha 1\",\n",
        "    \"EMILIN1\": \"elastin microfibril interfacer 1\",\n",
        "    \"EZR\": \"ezrin\",\n",
        "    \"F10\": \"coagulation factor X\",\n",
        "    \"F2\": \"coagulation factor II (thrombin)\",\n",
        "    \"F7\": \"coagulation factor VII\",\n",
        "    \"F9\": \"coagulation factor IX\",\n",
        "    \"FGL2\": \"fibrinogen-like 2\",\n",
        "    \"FMOD\": \"fibromodulin\",\n",
        "    \"FN1\": \"fibronectin 1\",\n",
        "    \"GAPDH\": \"glyceraldehyde-3-phosphate dehydrogenase\",\n",
        "    \"GAS6\": \"growth arrest-specific 6\",\n",
        "    \"GC\": \"group-specific component (vitamin D binding protein)\",\n",
        "    \"HAPLN3\": \"hyaluronan and proteoglycan link protein 3\",\n",
        "    \"HSP90B1\": \"heat shock protein 90 beta family member 1\",\n",
        "    \"HSPA5\": \"heat shock protein family A (Hsp70) member 5\",\n",
        "    \"HTRA1\": \"HtrA serine peptidase 1\",\n",
        "    \"IBSP\": \"integrin binding sialoprotein\",\n",
        "    \"IGF1\": \"insulin-like growth factor 1\",\n",
        "    \"IGF2\": \"insulin-like growth factor 2\",\n",
        "    \"IGFALS\": \"insulin-like growth factor binding protein, acid labile subunit\",\n",
        "    \"IGFBP1\": \"insulin-like growth factor binding protein 1\",\n",
        "    \"KNG1\": \"kininogen 1\",\n",
        "    \"KRT2\": \"keratin 2\",\n",
        "    \"LOX\": \"lysyl oxidase\",\n",
        "    \"LRRC15\": \"leucine rich repeat containing 15\",\n",
        "    \"LUM\": \"lumican\",\n",
        "    \"MGP\": \"matrix Gla protein\",\n",
        "    \"MMP2\": \"matrix metallopeptidase 2\",\n",
        "    \"MSN\": \"moesin\",\n",
        "    \"MYO1B\": \"myosin IB\",\n",
        "    \"NUCB1\": \"nucleobindin 1\",\n",
        "    \"NUCB2\": \"nucleobindin 2\",\n",
        "    \"OGN\": \"osteoglycin\",\n",
        "    \"OLFML1\": \"olfactomedin-like 1\",\n",
        "    \"OLFML3\": \"olfactomedin-like 3\",\n",
        "    \"OMD\": \"osteomodulin\",\n",
        "    \"P4HB\": \"prolyl 4-hydroxylase subunit beta\",\n",
        "    \"PAM\": \"peptidylglycine alpha-amidating monooxygenase\",\n",
        "    \"PCOLCE\": \"procollagen C-endopeptidase enhancer\",\n",
        "    \"PHOSPHO1\": \"phosphatase, orphan 1\",\n",
        "    \"POSTN\": \"periostin\",\n",
        "    \"PROC\": \"protein C, inactivator of coagulation factors Va and VIIIa\",\n",
        "    \"PROS1\": \"protein S, vitamin K-dependent\",\n",
        "    \"PRSS2\": \"protease, serine 2\",\n",
        "    \"SERPINC1\": \"serpin family C member 1 (antithrombin)\",\n",
        "    \"SERPIND1\": \"serpin family D member 1 (heparin cofactor II)\",\n",
        "    \"SERPINF1\": \"serpin family F member 1\",\n",
        "    \"SLC8A3\": \"solute carrier family 8 member A3\",\n",
        "    \"SPARC\": \"secreted protein acidic and rich in cysteine (osteonectin)\",\n",
        "    \"SPARCL1\": \"SPARC-like 1 (hevin)\",\n",
        "    \"SPP1\": \"secreted phosphoprotein 1 (osteopontin)\",\n",
        "    \"SPP2\": \"secreted phosphoprotein 2\",\n",
        "    \"TGFB1\": \"transforming growth factor beta 1\",\n",
        "    \"THBS1\": \"thrombospondin 1\",\n",
        "    \"TNC\": \"tenascin C\",\n",
        "    \"TPP1\": \"tripeptidyl peptidase I\",\n",
        "    \"TUBA1B\": \"tubulin alpha 1b\",\n",
        "    \"VCAN\": \"versican\",\n",
        "    \"VIT\": \"vitrin\",\n",
        "    \"VTN\": \"vitronectin\",\n",
        "    \"BANP\": \"BTG3-associated nuclear protein\",\n",
        "    \"BCL9L\": \"B-cell CLL/lymphoma 9-like protein\",\n",
        "    \"CD9\": \"CD9 molecule\",\n",
        "    \"CD36\": \"CD36 molecule (thrombospondin receptor)\",\n",
        "    \"CD63\": \"CD63 molecule\",\n",
        "    \"CD81\": \"CD81 molecule\",\n",
        "    \"COL4A1\": \"collagen type IV alpha 1 chain\",\n",
        "    \"COL4A2\": \"collagen type IV alpha 2 chain\",\n",
        "    \"COL6A5\": \"collagen type VI alpha 5 chain\",\n",
        "    \"COL8A2\": \"collagen type VIII alpha 2 chain\",\n",
        "    \"COL9A1\": \"collagen type IX alpha 1 chain\",\n",
        "    \"COL9A2\": \"collagen type IX alpha 2 chain\",\n",
        "    \"COL17A1\": \"collagen type XVII alpha 1 chain\",\n",
        "    \"COL18A1\": \"collagen type XVIII alpha 1 chain\",\n",
        "    \"COL26A1\": \"collagen type XXVI alpha 1 chain\",\n",
        "    \"COL28A1\": \"collagen type XXVIII alpha 1 chain\",\n",
        "    \"EMID1\": \"EMI domain containing 1\",\n",
        "    \"IGHA2\": \"immunoglobulin heavy constant alpha 2\",\n",
        "    \"LTF\": \"lactotransferrin\",\n",
        "    \"LYZ\": \"lysozyme\",\n",
        "    \"PIP\": \"prolactin-induced protein\",\n",
        "    \"S100A7A\": \"S100 calcium binding protein A7A\",\n",
        "    \"S100A9\": \"S100 calcium binding protein A9\",\n",
        "    \"TLR3\": \"toll like receptor 3\",\n",
        "    \"TMEM119\": \"transmembrane protein 119\"\n",
        "}"
      ],
      "metadata": {
        "id": "s7rgxQvHyHTg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "enamel_proteins = {\n",
        "    \"AHSG\": \"alpha-2-HS-glycoprotein (fetuin-A)\",\n",
        "    \"ALB\": \"serum albumin\",\n",
        "    \"AMBN\": \"ameloblastin\",\n",
        "    \"AMEL\": \"amelogenin (unspecified isoform)\",\n",
        "    \"AMELY\": \"amelogenin Y-linked\",\n",
        "    \"AMELX\": \"amelogenin X-linked\",\n",
        "    \"AMTN\": \"amelotin\",\n",
        "    \"COL1A1\": \"collagen type I alpha 1 chain\",\n",
        "    \"COL1A2\": \"collagen type I alpha 2 chain\",\n",
        "    \"COL17A1\": \"collagen type XVII alpha 1 chain\",\n",
        "    \"COL2A1\": \"collagen type II alpha 1 chain\",\n",
        "    \"DSPP\": \"dentin sialophosphoprotein\",\n",
        "    \"ENAM\": \"enamelin\",\n",
        "    \"KLK4\": \"kallikrein related peptidase 4\",\n",
        "    \"MMP20\": \"matrix metallopeptidase 20\",\n",
        "    \"ODAM\": \"odontogenic, ameloblast associated\",\n",
        "    \"SERPINC1\": \"serpin family C member 1 (antithrombin)\",\n",
        "    \"TUFT1\": \"tuftelin 1\"\n",
        "}\n",
        "\n",
        "food_crust_proteins = [\n",
        "    (\"CSN1S1\", \"alpha-S1 casein\"),\n",
        "    (\"CSN1S2\", \"alpha-S2 casein\"),\n",
        "    (\"CSN2\", \"beta-casein\"),\n",
        "    (\"CSN3\", \"kappa-casein\"),\n",
        "    (\"LALBA\", \"alpha-lactalbumin\"),\n",
        "    (\"LGB\", \"beta-lactoglobulin\"),\n",
        "    (\"IGH@\", \"immunoglobulin heavy chains\"),\n",
        "    (\"IGK@\", \"immunoglobulin kappa light chains\"),\n",
        "    (\"IGL@\", \"immunoglobulin lambda light chains\"),\n",
        "    (\"LTF\", \"lactoferrin\"),\n",
        "    (\"LYZ\", \"lysozyme\"),\n",
        "    (\"ALB\", \"serum albumin\"),\n",
        "    (\"MUC1\", \"mucin 1\"),\n",
        "    (\"BTN1A1\", \"butyrophilin subfamily 1 member A1\"),\n",
        "    (\"XDH\", \"xanthine dehydrogenase/oxidase\"),\n",
        "    (\"MFGE8\", \"milk fat globule-EGF factor 8 protein\")\n",
        "]\n",
        "\n",
        "saliva_proteins = [\n",
        "    # Digestive Enzymes\n",
        "    (\"AMY1\", \"Alpha-amylase\"),\n",
        "    (\"LING1\", \"Lingual lipase\"),\n",
        "    # Antimicrobial Proteins\n",
        "    (\"LYZ\", \"Lysozyme\"),\n",
        "    (\"LPO\", \"Lactoperoxidase\"),\n",
        "    (\"MUC7\", \"Mucin-7\"),\n",
        "    (\"MUC5B\", \"Mucin-5B\"),\n",
        "    (\"DMBT1\", \"Deleted in malignant brain tumors 1\"),\n",
        "    (\"S100A7\", \"Psoriasin\"),\n",
        "    (\"S100A8\", \"Calprotectin\"),\n",
        "    (\"S100A9\", \"Calprotectin\"),\n",
        "    (\"SLPI\", \"Secretory leukocyte protease inhibitor\"),\n",
        "    (\"BPIFA1\", \"BPI fold-containing family A member 1\"),\n",
        "    (\"BPIFB2\", \"BPI fold-containing family B member 2\"),\n",
        "    # Immunoglobulin and Immune System-Related Proteins\n",
        "    (\"IGHG1\", \"Immunoglobulin heavy constant gamma 1\"),\n",
        "    (\"IGHG2\", \"Immunoglobulin heavy constant gamma 2\"),\n",
        "    (\"IGHG3\", \"Immunoglobulin heavy constant gamma 3\"),\n",
        "    (\"IGHG4\", \"Immunoglobulin heavy constant gamma 4\"),\n",
        "    (\"IGHA1\", \"Immunoglobulin heavy constant alpha 1\"),\n",
        "    (\"IGHA2\", \"Immunoglobulin heavy constant alpha 2\"),\n",
        "    (\"IGKC\", \"Immunoglobulin kappa constant\"),\n",
        "    (\"IGLC1\", \"Immunoglobulin lambda constant 1\"),\n",
        "    (\"IGLC2\", \"Immunoglobulin lambda constant 2\"),\n",
        "    (\"IGLC3\", \"Immunoglobulin lambda constant 3\")\n",
        "]\n",
        "\n",
        "milk_proteins = [\n",
        "    (\"CSN1S1\", \"alpha-S1-casein\"),\n",
        "    (\"CSN2\", \"beta-casein\"),\n",
        "    (\"CSN3\", \"kappa-casein\"),\n",
        "    (\"CSN1S2\", \"alpha-S2-casein\"),\n",
        "    (\"LALBA\", \"alpha-lactalbumin\"),\n",
        "    (\"BLG\", \"beta-lactoglobulin\"),\n",
        "    (\"PAEP\", \"Pregnancy-associated glycoprotein\"),\n",
        "    (\"LCN2\", \"Lactoferrin\"),\n",
        "    (\"LTF\", \"Lactotransferrin\"),\n",
        "    (\"ALB\", \"Serum albumin\"),\n",
        "    (\"HP\", \"Haptoglobin\"),\n",
        "    (\"IGLC1\", \"Immunoglobulin lambda constant 1\"),\n",
        "    (\"IGLC2\", \"Immunoglobulin lambda constant 2\"),\n",
        "    (\"IGLC3\", \"Immunoglobulin lambda constant 3\"),\n",
        "    (\"IGHA1\", \"Immunoglobulin alpha 1\"),\n",
        "    (\"IGHA2\", \"Immunoglobulin alpha 2\"),\n",
        "    (\"IGHG1\", \"Immunoglobulin gamma 1\"),\n",
        "    (\"IGHG2\", \"Immunoglobulin gamma 2\"),\n",
        "    (\"IGHG3\", \"Immunoglobulin gamma 3\"),\n",
        "    (\"IGHG4\", \"Immunoglobulin gamma 4\"),\n",
        "    (\"MFGE8\", \"Milk fat globule-EGF factor 8\"),\n",
        "    (\"BTN1A1\", \"Butyrophilin subfamily 1 member A1\"),\n",
        "    (\"XDH\", \"Xanthine dehydrogenase\"),\n",
        "    (\"AZGP1\", \"Zinc-alpha-2-glycoprotein\")\n",
        "]\n",
        "\n",
        "muscle_proteins = [\n",
        "    (\"ACTB\", \"Actin, cytoplasmic 1\"),\n",
        "    (\"ATP2A1\", \"Sarcoplasmic/endoplasmic reticulum calcium ATPase 1\"),\n",
        "    (\"ATP5F1B\", \"ATP synthase subunit beta, mitochondrial precursor\"),\n",
        "    (\"ATP6\", \"ATP synthase subunit a\"),\n",
        "    (\"CACNA1S\", \"Calcium Voltage-Gated Channel Subunit Alpha1 S\"),\n",
        "    (\"CAMK2G\", \"Calcium/Calmodulin Dependent Protein Kinase II Gamma\"),\n",
        "    (\"CKMT2\", \"Creatine kinase, testis isozyme\"),\n",
        "    (\"COL1A1\", \"Collagen type I alpha 1 chain\"),\n",
        "    (\"COL1A2\", \"Collagen type I alpha 2 chain\"),\n",
        "    (\"COL2A1\", \"Collagen type II alpha 1 chain\"),\n",
        "    (\"COL3A1\", \"Collagen type III alpha 1 chain\"),\n",
        "    (\"COL5A1\", \"Collagen type V alpha 1 chain\"),\n",
        "    (\"COL5A2\", \"Collagen type V alpha 2 chain\"),\n",
        "    (\"CS\", \"Citrate synthase, mitochondrial precursor\"),\n",
        "    (\"ELOVL5\", \"ELOVL Fatty Acid Elongase 5\"),\n",
        "    (\"FABP3\", \"Fatty Acid Binding Protein 3\"),\n",
        "    (\"FABP4\", \"Fatty Acid Binding Protein 4\"),\n",
        "    (\"FASN\", \"Fatty Acid Synthase\"),\n",
        "    (\"HBA\", \"Hemoglobin cathodic subunit alpha\"),\n",
        "    (\"HBB\", \"Hemoglobin subunit beta\"),\n",
        "    (\"LDHA\", \"L-lactate dehydrogenase A chain\"),\n",
        "    (\"LDHB\", \"L-lactate dehydrogenase B chain\"),\n",
        "    (\"MB\", \"Myoglobin\"),\n",
        "    (\"MC4R\", \"Melanocortin 4 Receptor\"),\n",
        "    (\"MYH4\", \"Myosin heavy chain, fast skeletal muscle\"),\n",
        "    (\"MYH6\", \"Atrial myosin heavy chain\"),\n",
        "    (\"MYH9A\", \"Myosin-9a\"),\n",
        "    (\"MYH11A\", \"Myosin-11a\"),\n",
        "    (\"MYLPFA\", \"Myosin regulatory light chain 2, skeletal muscle isoform A\"),\n",
        "    (\"PPP1R12A\", \"Protein phosphatase 1 regulatory subunit 12A\"),\n",
        "    (\"MYLK3\", \"Myosin light chain kinase 3\"),\n",
        "    (\"MYO6A\", \"Unconventional myosin-VI\"),\n",
        "    (\"TPI1\", \"Triosephosphate isomerase\"),\n",
        "    (\"TPM1\", \"Tropomyosin-1 alpha chain\"),\n",
        "    (\"MYH9B\", \"Myosin-9b\"),\n",
        "    (\"MYL1\", \"Myosin light chain 3, skeletal muscle isoform\"),\n",
        "    (\"MYLIPA\", \"E3 ubiquitin-protein ligase MYLIP-A\"),\n",
        "    (\"MYO1C\", \"Unconventional myosin-Ic\"),\n",
        "    (\"MYO1D\", \"Unconventional myosin-Id\"),\n",
        "    (\"MYO5B\", \"Myosin VB\"),\n",
        "    (\"MYO6B\", \"Unconventional myosin-VI\"),\n",
        "    (\"MYO9AA\", \"Unconventional myosin-IXAa\"),\n",
        "    (\"MYO9AB\", \"Unconventional myosin-IXAb\"),\n",
        "    (\"NDPKB\", \"Nucleoside diphosphate kinase B\"),\n",
        "    (\"PDE3B\", \"Phosphodiesterase 3B\"),\n",
        "    (\"PRKAG3\", \"Protein Kinase AMP-Activated Non-Catalytic Subunit Gamma 3\"),\n",
        "    (\"PVALB\", \"Parvalbumin beta\"),\n",
        "    (\"PVALB2\", \"Parvalbumin-2\"),\n",
        "    (\"PVALB7\", \"Parvalbumin-7\"),\n",
        "    (\"PVALB8\", \"Parvalbumin-8\"),\n",
        "    (\"PVALB9\", \"Parvalbumin-9\"),\n",
        "    (\"RYR1\", \"Ryanodine Receptor 1\")\n",
        "]\n",
        "\n",
        "blood_proteins = [\n",
        "    (\"ALB\", \"Albumin\"),\n",
        "    (\"HBA\", \"Hemoglobin subunit alpha\"),\n",
        "    (\"HBB\", \"Hemoglobin subunit beta\"),\n",
        "    (\"TTR\", \"Transthyretin\"),\n",
        "    (\"APOA1\", \"Apolipoprotein A-I\"),\n",
        "    (\"APOB\", \"Apolipoprotein B\"),\n",
        "    (\"APOC3\", \"Apolipoprotein C-III\"),\n",
        "    (\"TF\", \"Transferrin\"),\n",
        "    (\"IGG1\", \"Immunoglobulin G1\"),\n",
        "    (\"IGG2\", \"Immunoglobulin G2\"),\n",
        "    (\"IGG3\", \"Immunoglobulin G3\"),\n",
        "    (\"IGG4\", \"Immunoglobulin G4\"),\n",
        "    (\"IGA1\", \"Immunoglobulin A1\"),\n",
        "    (\"IGA2\", \"Immunoglobulin A2\"),\n",
        "    (\"IGM\", \"Immunoglobulin M\"),\n",
        "    (\"FGA\", \"Fibrinogen alpha chain\"),\n",
        "    (\"FGB\", \"Fibrinogen beta chain\"),\n",
        "    (\"FGG\", \"Fibrinogen gamma chain\"),\n",
        "    (\"CRP\", \"C-reactive protein\"),\n",
        "    (\"CP\", \"Ceruloplasmin\"),\n",
        "    (\"HP\", \"Haptoglobin\"),\n",
        "    (\"GC\", \"Vitamin D-binding protein\"),\n",
        "    (\"AHSG\", \"Alpha-2-HS-glycoprotein\"),\n",
        "    (\"C3\", \"Complement C3\"),\n",
        "    (\"C4A\", \"Complement C4A\"),\n",
        "    (\"C4B\", \"Complement C4B\"),\n",
        "    (\"C5\", \"Complement C5\"),\n",
        "    (\"SAA1\", \"Serum amyloid A1\"),\n",
        "    (\"SAA2\", \"Serum amyloid A2\"),\n",
        "    (\"KLK3\", \"Prostate-specific antigen\"),\n",
        "    (\"AMBP\", \"Alpha-1-microglobulin/bikunin precursor\"),\n",
        "    (\"ORM1\", \"Orosomucoid 1\"),\n",
        "    (\"ORM2\", \"Orosomucoid 2\"),\n",
        "    (\"SELM\", \"Selenoprotein M\"),\n",
        "    (\"SELP\", \"P-selectin\"),\n",
        "    (\"SELE\", \"E-selectin\"),\n",
        "    (\"SELL\", \"L-selectin\"),\n",
        "    (\"VWF\", \"Von Willebrand factor\"),\n",
        "    (\"PLG\", \"Plasminogen\"),\n",
        "    (\"F12\", \"Coagulation factor XII\"),\n",
        "    (\"F11\", \"Coagulation factor XI\"),\n",
        "    (\"F9\", \"Coagulation factor IX\"),\n",
        "    (\"F8\", \"Coagulation factor VIII\"),\n",
        "    (\"F7\", \"Coagulation factor VII\"),\n",
        "    (\"F10\", \"Coagulation factor X\"),\n",
        "    (\"F5\", \"Coagulation factor V\"),\n",
        "    (\"PROC\", \"Protein C\"),\n",
        "    (\"PROS1\", \"Protein S\"),\n",
        "    (\"THBD\", \"Thrombomodulin\"),\n",
        "    (\"ATF3\", \"Activating transcription factor 3\"),\n",
        "    (\"SERPINA1\", \"Alpha-1-antitrypsin\"),\n",
        "    (\"SERPINA3\", \"Alpha-1-antichymotrypsin\")\n",
        "]\n",
        "\n",
        "skin_keratins =  [\n",
        "        (\"KRT1\", \"Keratin, type II cytoskeletal 1\"),\n",
        "        (\"KRT2\", \"Keratin, type II cytoskeletal 2\"),\n",
        "        (\"KRT3\", \"Keratin, type II cytoskeletal 3\"),\n",
        "        (\"KRT4\", \"Keratin, type II cytoskeletal 4\"),\n",
        "        (\"KRT5\", \"Keratin, type II cytoskeletal 5\"),\n",
        "        (\"KRT6A\", \"Keratin, type II cytoskeletal 6A\"),\n",
        "        (\"KRT6B\", \"Keratin, type II cytoskeletal 6B\"),\n",
        "        (\"KRT6C\", \"Keratin, type II cytoskeletal 6C\"),\n",
        "        (\"KRT9\", \"Keratin, type I cytoskeletal 9\"),\n",
        "        (\"KRT10\", \"Keratin, type I cytoskeletal 10\"),\n",
        "        (\"KRT13\", \"Keratin, type I cytoskeletal 13\"),\n",
        "        (\"KRT14\", \"Keratin, type I cytoskeletal 14\"),\n",
        "        (\"KRT15\", \"Keratin, type I cytoskeletal 15\"),\n",
        "        (\"KRT16\", \"Keratin, type I cytoskeletal 16\"),\n",
        "        (\"KRT24\", \"Keratin, type I cytoskeletal 24\"),\n",
        "        (\"KRT31\", \"Keratin, type I cytoskeletal 31\"),\n",
        "        (\"KRT32\", \"Keratin, type I cytoskeletal 32\"),\n",
        "        (\"KRT33B\", \"Keratin, type I cytoskeletal 33B\"),\n",
        "        (\"KRT34\", \"Keratin, type I cytoskeletal 34\"),\n",
        "        (\"KRT37\", \"Keratin, type I cytoskeletal 37\"),\n",
        "        (\"KRT38\", \"Keratin, type I cytoskeletal 38\"),\n",
        "        (\"KRT74\", \"Keratin, type II cytoskeletal 74\"),\n",
        "        (\"KRT77\", \"Keratin, type II cytoskeletal 77\"),\n",
        "        (\"KRT78\", \"Keratin, type II cytoskeletal 78\"),\n",
        "        (\"KRT82\", \"Keratin, type II cytoskeletal 82\"),\n",
        "        (\"KRT83\", \"Keratin, type II cytoskeletal 83\"),\n",
        "        (\"KRT85\", \"Keratin, type II cytoskeletal 85\"),\n",
        "        (\"KRT86\", \"Keratin, type II cytoskeletal 86\")\n",
        "]\n",
        "\n",
        "hair_keratins  =  [\n",
        "        (\"KRT31\", \"Keratin, type I cytoskeletal 31\"),\n",
        "        (\"KRT32\", \"Keratin, type I cytoskeletal 32\"),\n",
        "        (\"KRT33A\", \"Keratin, type I cytoskeletal 33A\"),\n",
        "        (\"KRT33B\", \"Keratin, type I cytoskeletal 33B\"),\n",
        "        (\"KRT34\", \"Keratin, type I cytoskeletal 34\"),\n",
        "        (\"KRT35\", \"Keratin, type I cytoskeletal 35\"),\n",
        "        (\"KRT36\", \"Keratin, type I cytoskeletal 36\"),\n",
        "        (\"KRT37\", \"Keratin, type I cytoskeletal 37\"),\n",
        "        (\"KRT38\", \"Keratin, type I cytoskeletal 38\"),\n",
        "        (\"KRT39\", \"Keratin, type I cytoskeletal 39\"),\n",
        "        (\"KRT40\", \"Keratin, type I cytoskeletal 40\"),\n",
        "        (\"KRT81\", \"Keratin, type II cytoskeletal 81\"),\n",
        "        (\"KRT83\", \"Keratin, type II cytoskeletal 83\"),\n",
        "        (\"KRT85\", \"Keratin, type II cytoskeletal 85\"),\n",
        "        (\"KRT86\", \"Keratin, type II cytoskeletal 86\"),\n",
        "        # High sulfur proteins\"\n",
        "        (\"KAP1\", \"Keratin-associated protein 1\"),\n",
        "        (\"KAP2\", \"Keratin-associated protein 2\"),\n",
        "        (\"KAP3\", \"Keratin-associated protein 3\"),\n",
        "        (\"KAP4\", \"Keratin-associated protein 4\"),\n",
        "        (\"KAP5\", \"Keratin-associated protein 5\"),\n",
        "        (\"KAP6\", \"Keratin-associated protein 6\"),\n",
        "        (\"KAP7\", \"Keratin-associated protein 7\"),\n",
        "        (\"KAP8\", \"Keratin-associated protein 8\"),\n",
        "        (\"KAP9\", \"Keratin-associated protein 9\"),\n",
        "        (\"KAP10\", \"Keratin-associated protein 10\"),\n",
        "        (\"KAP11\", \"Keratin-associated protein 11\"),\n",
        "        (\"KAP12\", \"Keratin-associated protein 12\"),\n",
        "        (\"KAP13\", \"Keratin-associated protein 13\"),\n",
        "        (\"KAP14\", \"Keratin-associated protein 14\"),\n",
        "        (\"KAP15\", \"Keratin-associated protein 15\"),\n",
        "        (\"KAP16\", \"Keratin-associated protein 16\"),\n",
        "        (\"KAP17\", \"Keratin-associated protein 17\"),\n",
        "        (\"KAP18\", \"Keratin-associated protein 18\"),\n",
        "        (\"KAP19\", \"Keratin-associated protein 19\"),\n",
        "        (\"KAP20\", \"Keratin-associated protein 20\")\n",
        "]\n",
        "\n",
        "\n",
        "zooms_genes = [\n",
        "    (\"COL1A2\", \"collagen type I alpha 2 chain\"),\n",
        "    (\"COL1A1\", \"collagen type I alpha 1 chain\"),\n",
        "    (\"COL3A1\", \"collagen type III alpha 1 chain\")\n",
        "]\n",
        "\n",
        "# Define a combined dictionary of all genes\n",
        "all_genes = {\n",
        "    \"Bone\": bone_proteins,\n",
        "    \"Enamel\": enamel_proteins,\n",
        "    \"Food Crust\": food_crust_proteins,\n",
        "    \"Muscle\": muscle_proteins,\n",
        "    \"Blood\": blood_proteins,\n",
        "    \"Skin\": skin_keratins,\n",
        "    \"Hair\": hair_keratins,\n",
        "    \"Saliva\": saliva_proteins,\n",
        "    \"Milk\": milk_proteins,\n",
        "    \"Zooms\": zooms_genes\n",
        "}\n",
        "\n",
        "# Define the taxonomy IDs and their names\n",
        "taxonomy_dict = {\n",
        "    # Carnivora\n",
        "    33554: \"Carnivora\", 379584: \"Caniformia\", 379583: \"Feliformia\", 9632: \"Ursidae\", 9615: \"Canis lupus familiaris\", 9647: \"Procyonidae\",\n",
        "\n",
        "    # Bovidae\n",
        "    35500: \"Pecora\", 9850: \"Cervidae\", 9913: \"Bovidae\", 9895: \"Bos taurus\", 9925: \"Capra hircus\", 9940: \"Ovis aries\",\n",
        "\n",
        "    # Lagomorpha\n",
        "    9975: \"Lagomorpha\",\n",
        "\n",
        "    # Humans\n",
        "    9606: \"Homo sapiens\", 9605: \"Homo\", 9604: \"Hominidae\", 9443: \"Primates\", 40674: \"Mammalia\", 7711: \"Chordata\", 33208: \"Animalia\",\n",
        "\n",
        "    # Ostriches\n",
        "    8839: \"Struthio camelus\", 8838: \"Struthio\", 8837: \"Struthionidae\", 30498: \"Struthioniformes\", 8782: \"Aves\",\n",
        "\n",
        "    # Cod\n",
        "    8049: \"Gadus morhua\", 8048: \"Gadus\", 27502: \"Gadidae\", 27682: \"Gadiformes\", 7898: \"Actinopterygii\",\n",
        "\n",
        "    # Geckos\n",
        "    103697: \"Gekko gecko\", 8489: \"Gekko\", 8483: \"Gekkonidae\", 8509: \"Squamata\", 8457: \"Reptilia\"\n",
        "}\n",
        "\n",
        "# print(\"all_genes:\")\n",
        "# for tissue, genes in all_genes.items():\n",
        "#     print(f\"  {tissue}: {len(genes)} genes\")\n",
        "\n",
        "# print(\"\\ntaxonomy_dict:\")\n",
        "# for tax_id, tax_name in taxonomy_dict.items():\n",
        "#     print(f\"  {tax_id}: {tax_name}\")"
      ],
      "metadata": {
        "id": "v-hYV_bKsdqm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title #Functions\n",
        "\n",
        "def api_taxid_descendent_list(tax_id):\n",
        "    \"\"\"Gets all descendant taxonomy IDs for a given tax_id\"\"\"\n",
        "    print(f\"\\nGetting taxonomy descendants for TaxID: {tax_id}\")\n",
        "    url_tax_id_descendent = f\"https://rest.uniprot.org/taxonomy/stream?format=list&query=%28%28ancestor%3A{str(tax_id)}%29%29\"\n",
        "    tax_id_download = str(requests.get(url_tax_id_descendent).text)\n",
        "    tax_id_descendent_list = tax_id_download.split(\"\\n\")\n",
        "    tax_id_descendent_list.pop()  # Remove empty last element\n",
        "    tax_id_descendent_list = [int(taxid) for taxid in tax_id_descendent_list]\n",
        "    tax_id_descendent_list.append(int(tax_id))\n",
        "    return tax_id_descendent_list\n",
        "\n",
        "def api_get_json_batch(batch_url):\n",
        "    \"\"\"Get JSON results in batches of 500\"\"\"\n",
        "    re_next_link = re.compile(r'<(.+)>; rel=\"next\"')\n",
        "    retries = Retry(total=5, backoff_factor=0.25, status_forcelist=[500, 502, 503, 504])\n",
        "    session = requests.Session()\n",
        "    session.mount(\"https://\", HTTPAdapter(max_retries=retries))\n",
        "\n",
        "    while batch_url:\n",
        "        records_batch = session.get(batch_url)\n",
        "        records_batch.raise_for_status()\n",
        "        records_batch_json = records_batch.json()\n",
        "        record_count = int(records_batch.headers[\"x-total-results\"])\n",
        "\n",
        "        yield records_batch_json, record_count\n",
        "\n",
        "        if \"Link\" in records_batch.headers:\n",
        "            match = re_next_link.match(records_batch.headers[\"Link\"])\n",
        "            batch_url = match.group(1)\n",
        "        else:\n",
        "            batch_url = None\n",
        "\n",
        "def json_to_fasta(protein_record, tax_id_descendent_list, gene_name=None):\n",
        "    \"\"\"Convert JSON record to FASTA format\"\"\"\n",
        "    # Check for required fields\n",
        "    if 'sequence' not in protein_record or 'value' not in protein_record['sequence']:\n",
        "        return None, None\n",
        "\n",
        "    header = f\"UPI|{protein_record['uniParcId']}\"\n",
        "    if gene_name:\n",
        "        header += f\"|{gene_name}\"\n",
        "\n",
        "    # Add UniProt accessions if available\n",
        "    if 'uniProtKBAccessions' in protein_record and protein_record['uniProtKBAccessions']:\n",
        "        header += f\"|{protein_record['uniProtKBAccessions'][0]}\"\n",
        "\n",
        "    # Add creation date if available\n",
        "    if 'oldestCrossRefCreated' in protein_record:\n",
        "        header += f\"|{protein_record['oldestCrossRefCreated']}\"\n",
        "\n",
        "    # Add gene name\n",
        "    if gene_name:\n",
        "        header += f\" GN={gene_name}\"\n",
        "\n",
        "    sequence = protein_record['sequence']['value']\n",
        "    return header, sequence\n",
        "\n",
        "def download_proteins(tax_id, gene_list, raw_fasta):\n",
        "    \"\"\"Download proteins for given taxonomy ID and gene list\"\"\"\n",
        "    print(f\"\\nStarting protein download with TaxID: {tax_id}\")\n",
        "    tax_id_descendent_list = api_taxid_descendent_list(tax_id)\n",
        "    records_fasta_list = []\n",
        "    total_count = 0\n",
        "\n",
        "    for gene in gene_list:\n",
        "        # First try exact taxonomy match\n",
        "        url = f\"https://rest.uniprot.org/uniparc/search?compressed=false&format=json&query=%28%28gene%3A{gene}%29%20AND%20%28taxonomy_id%3A{tax_id}%29%20NOT%20%28database%3AFusionGDB%29%29&size=500\"\n",
        "        print(f\"\\nRequesting URL: {url}\")\n",
        "\n",
        "        # Make initial request and print response info\n",
        "        initial_response = requests.get(url)\n",
        "        print(f\"Response status code: {initial_response.status_code}\")\n",
        "        initial_data = initial_response.json()\n",
        "        results = initial_data.get('results', [])\n",
        "        print(f\"Found {len(results)} results\")\n",
        "\n",
        "        # If no results, try with parent taxonomy (Homo for Homo sapiens)\n",
        "        if not results:\n",
        "            parent_taxid = 9605  # Homo genus\n",
        "            url = f\"https://rest.uniprot.org/uniparc/search?compressed=false&format=json&query=%28%28gene%3A{gene}%29%20AND%20%28taxonomy_id%3A{parent_taxid}%29%20NOT%20%28database%3AFusionGDB%29%29&size=500\"\n",
        "            print(f\"\\nNo results with species taxid, trying genus taxid: {parent_taxid}\")\n",
        "            print(f\"New URL: {url}\")\n",
        "            initial_response = requests.get(url)\n",
        "            initial_data = initial_response.json()\n",
        "            results = initial_data.get('results', [])\n",
        "\n",
        "        try:\n",
        "            for record in results:\n",
        "                header, sequence = json_to_fasta(record, tax_id_descendent_list, gene)\n",
        "                if header and sequence:\n",
        "                    record_fasta = SeqRecord(Seq(sequence),\n",
        "                                           id=header.replace(\",\", \"\").replace(\";\", \"\").replace(\":\", \"\"),\n",
        "                                           description=\"\")\n",
        "                    records_fasta_list.append(record_fasta)\n",
        "                    total_count += 1\n",
        "            print(f\"  Downloaded {total_count} proteins so far...\")\n",
        "\n",
        "        except requests.exceptions.RequestException as e:\n",
        "            print(f\"Error downloading {gene}: {str(e)}\")\n",
        "            continue\n",
        "\n",
        "    # Write results to file\n",
        "    if records_fasta_list:\n",
        "        with open(raw_fasta, \"w\") as output_handle:\n",
        "            SeqIO.write(records_fasta_list, output_handle, \"fasta\")\n",
        "        print(f\"Successfully downloaded {total_count} proteins\")\n",
        "        return total_count\n",
        "    return 0\n",
        "\n",
        "def remove_duplicates(raw_fasta, output_folder):\n",
        "    \"\"\"Remove duplicate sequences from the FASTA file\"\"\"\n",
        "    if not os.path.exists(raw_fasta):\n",
        "        return 0\n",
        "\n",
        "    records = list(SeqIO.parse(raw_fasta, \"fasta\"))\n",
        "    unique_sequences = {}\n",
        "\n",
        "    for record in records:\n",
        "        seq = str(record.seq)\n",
        "        if seq not in unique_sequences:\n",
        "            unique_sequences[seq] = record\n",
        "\n",
        "    # Create fasta_remove_redundancy folder\n",
        "    redundancy_folder = os.path.join(output_folder, \"fasta_remove_redundancy\")\n",
        "    os.makedirs(redundancy_folder, exist_ok=True)\n",
        "\n",
        "    # Save unfiltered database\n",
        "    shutil.copy(raw_fasta, os.path.join(redundancy_folder, \"unfiltered_database.fasta\"))\n",
        "\n",
        "    # Write unique sequences\n",
        "    filtered_path = os.path.join(redundancy_folder, \"filtered_database.fasta\")\n",
        "    with open(filtered_path, \"w\") as handle:\n",
        "        SeqIO.write(unique_sequences.values(), handle, \"fasta\")\n",
        "\n",
        "    # Move filtered database to main folder\n",
        "    output_path = os.path.join(output_folder, \"filtered_mf.fasta\")\n",
        "    shutil.copy(filtered_path, output_path)\n",
        "\n",
        "    return len(unique_sequences)\n",
        "\n",
        "def generate_metadata(filtered_fasta, metadata_folder, gene_list_path):\n",
        "    \"\"\"Generate metadata files about the database\"\"\"\n",
        "    os.makedirs(metadata_folder, exist_ok=True)\n",
        "\n",
        "    # Read the FASTA file\n",
        "    records = list(SeqIO.parse(filtered_fasta, \"fasta\"))\n",
        "\n",
        "    # Generate basic stats\n",
        "    with open(os.path.join(metadata_folder, \"summary.txt\"), \"w\") as f:\n",
        "        f.write(f\"Total sequences: {len(records)}\\n\")\n",
        "\n",
        "        if gene_list_path:\n",
        "            with open(gene_list_path) as gene_file:\n",
        "                genes = [line.strip() for line in gene_file if line.strip()]\n",
        "                f.write(f\"Total genes searched: {len(genes)}\\n\")\n",
        "\n",
        "                # Find which genes were found\n",
        "                found_genes = set()\n",
        "                for record in records:\n",
        "                    for gene in genes:\n",
        "                        if f\"GN={gene}\" in record.id:\n",
        "                            found_genes.add(gene)\n",
        "\n",
        "                f.write(f\"Genes found: {len(found_genes)}\\n\")\n",
        "\n",
        "                # Write genes not found\n",
        "                not_found = set(genes) - found_genes\n",
        "                if not_found:\n",
        "                    with open(os.path.join(metadata_folder, \"genes_NOT_retrieved.txt\"), \"w\") as nf:\n",
        "                        for gene in sorted(not_found):\n",
        "                            nf.write(f\"{gene}\\n\")\n",
        "\n",
        "    # Generate species statistics\n",
        "    species_data = {}\n",
        "    for record in records:\n",
        "        match = re.search(r\"OS=([^OX]+)OX=(\\d+)\", record.id)\n",
        "        if match:\n",
        "            species = match.group(1).strip()\n",
        "            taxid = match.group(2)\n",
        "            if species not in species_data:\n",
        "                species_data[species] = {\"count\": 0, \"taxid\": taxid}\n",
        "            species_data[species][\"count\"] += 1\n",
        "\n",
        "    with open(os.path.join(metadata_folder, \"species_retrieved.csv\"), \"w\") as f:\n",
        "        f.write(\"Species,TaxID,Count\\n\")\n",
        "        for species, data in species_data.items():\n",
        "            f.write(f\"{species},{data['taxid']},{data['count']}\\n\")\n",
        "\n",
        "def run_integrated_proteoparc(tax_id, gene_list_path, output_folder):\n",
        "    \"\"\"Run the complete pipeline\"\"\"\n",
        "    os.makedirs(output_folder, exist_ok=True)\n",
        "\n",
        "    # Read gene list\n",
        "    with open(gene_list_path) as f:\n",
        "        gene_list = [line.strip() for line in f if line.strip()]\n",
        "\n",
        "    # Step 1: Download proteins\n",
        "    raw_fasta = os.path.join(output_folder, \"raw_proteins.fasta\")\n",
        "    print(\"Downloading proteins...\")\n",
        "    num_downloaded = download_proteins(tax_id, gene_list, raw_fasta)\n",
        "\n",
        "    if num_downloaded == 0:\n",
        "        print(\"No proteins found\")\n",
        "        return\n",
        "\n",
        "    print(f\"Downloaded {num_downloaded} proteins\")\n",
        "\n",
        "    # Step 2: Remove duplicates\n",
        "    print(\"Removing duplicates...\")\n",
        "    num_unique = remove_duplicates(raw_fasta, output_folder)\n",
        "    print(f\"Reduced to {num_unique} unique sequences\")\n",
        "\n",
        "    # Step 3: Generate metadata\n",
        "    print(\"Generating metadata...\")\n",
        "    metadata_folder = os.path.join(output_folder, \"metadata\")\n",
        "    generate_metadata(os.path.join(output_folder, \"filtered_mf.fasta\"),\n",
        "                     metadata_folder,\n",
        "                     gene_list_path)\n",
        "\n",
        "    print(f\"Pipeline completed. Results in: {output_folder}\")\n",
        "\n",
        "def run_proteoparc(b):\n",
        "    \"\"\"Run the ProteoParc pipeline\"\"\"\n",
        "    if not hasattr(run_proteoparc, 'taxonomy_id'):\n",
        "        print(\"Please select a taxonomy first\")\n",
        "        return\n",
        "\n",
        "    paths = create_gene_list_file(b, taxonomy_dict, run_proteoparc.taxonomy_id) # Pass taxonomy_dict and taxonomy_id\n",
        "    if paths:\n",
        "        run_integrated_proteoparc(run_proteoparc.taxonomy_id, paths['genelist'], paths['root'])\n",
        "\n",
        "def get_selected_tissue():\n",
        "    \"\"\"Determines active tissue type from selected genes\"\"\"\n",
        "    selected_genes = get_selected_genes()\n",
        "    tissue_counts = {}\n",
        "\n",
        "    for gene in selected_genes:\n",
        "        for tissue, genes in all_genes.items():\n",
        "            if any(gene == g[0] for g in genes):\n",
        "                tissue_counts[tissue] = tissue_counts.get(tissue, 0) + 1\n",
        "\n",
        "    if not tissue_counts:\n",
        "        return \"mixed\"\n",
        "    return max(tissue_counts.items(), key=lambda x: x[1])[0]\n",
        "\n",
        "def create_project_structure(taxonomy_dict, taxonomy_id, base_path=\"/content/drive/MyDrive/Colab_Notebooks/palaeoparc\"):\n",
        "    \"\"\"\n",
        "    Creates directory structure:\n",
        "    base_path/\n",
        "        tissue_type/\n",
        "            taxa_name/\n",
        "                YYYYMMDD_HHMMSS/\n",
        "    \"\"\"\n",
        "    timestamp = datetime.now().strftime(\"%Y%m%d_%H%M%S\")\n",
        "    tissue_type = get_selected_tissue()\n",
        "    taxa_name = taxonomy_dict.get(taxonomy_id, \"unknown_taxonomy\")\n",
        "\n",
        "    # Sanitize names\n",
        "    tissue_folder = tissue_type.lower().replace(\" \", \"_\")\n",
        "    taxa_folder = taxa_name.lower().replace(\" \", \"_\")\n",
        "\n",
        "    # Build paths\n",
        "    timestamp_path = os.path.join(base_path, tissue_folder, taxa_folder, timestamp)\n",
        "    os.makedirs(timestamp_path, exist_ok=True)\n",
        "    os.makedirs(os.path.join(timestamp_path, \"metadata\"), exist_ok=True)\n",
        "\n",
        "    # Generate file paths\n",
        "    fasta_name = f\"{tissue_folder}_{taxa_folder}.fasta\"\n",
        "\n",
        "    return {\n",
        "        'root': timestamp_path,\n",
        "        'fasta': os.path.join(timestamp_path, fasta_name),\n",
        "        'genelist': os.path.join(timestamp_path, \"genelist.txt\"),\n",
        "        'metadata': os.path.join(timestamp_path, \"metadata\")\n",
        "    }\n"
      ],
      "metadata": {
        "id": "b0aZicxOluie",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title GUI\n",
        "def create_taxonomy_buttons(taxonomy_dict):\n",
        "    buttons = []\n",
        "    for tax_id, name in taxonomy_dict.items():\n",
        "        button = widgets.Button(\n",
        "            description=name,\n",
        "            tooltip=f\"{name} (ID: {tax_id})\",\n",
        "            layout=widgets.Layout(width='auto', height='auto'),\n",
        "            style={'button_color': '#e7e7e7', 'font_weight': 'normal'}\n",
        "        )\n",
        "\n",
        "        def on_button_click(b, tax_id=tax_id, name=name):\n",
        "            global selected_taxonomy_id\n",
        "            selected_taxonomy_id = tax_id\n",
        "            output_dir = create_output_directory(taxonomy_dict, tax_id)\n",
        "            create_gene_list_file.output_dir = output_dir  # Store output_dir for later use\n",
        "            run_proteoparc.taxonomy_id = tax_id  # Store taxonomy_id for later use\n",
        "            url = f\"https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id={tax_id}\"\n",
        "            clear_output()\n",
        "            display(HTML(\"<h2>Selected Taxonomy</h2>\"))\n",
        "            display(HTML(f\"<h3>{name} (ID: {tax_id})</h3>\"))\n",
        "            display(HTML(f\"<a href='{url}' target='_blank'>Open NCBI Taxonomy Browser</a>\"))\n",
        "            display(HTML(f\"<p>Output directory: {output_dir}</p>\"))\n",
        "            display(HTML(\"<h2>Select Taxonomy</h2>\"))\n",
        "            display(button_grid)\n",
        "            display(HTML(\"<h2>Gene Selection by Tissue</h2>\"))\n",
        "            display(widgets.HBox([select_all_button, deselect_all_button]))\n",
        "            display(tissue_selection)\n",
        "            display(widgets.HBox([print_button, save_button, run_proteoparc_button]))\n",
        "\n",
        "        button.on_click(on_button_click)\n",
        "        buttons.append(button)\n",
        "\n",
        "    button_grid = widgets.GridBox(\n",
        "        buttons,\n",
        "        layout=widgets.Layout(\n",
        "            grid_template_columns='repeat(3, auto)',\n",
        "            grid_gap='10px',\n",
        "            width='100%'\n",
        "        )\n",
        "    )\n",
        "\n",
        "    return button_grid\n",
        "\n",
        "def create_gene_checkboxes(gene_list):\n",
        "    return [widgets.Checkbox(\n",
        "        value=False,\n",
        "        description=f\"{gene} ({desc})\",\n",
        "        indent=False,\n",
        "        layout=widgets.Layout(width='auto')\n",
        "    ) for gene, desc in gene_list]\n",
        "\n",
        "\n",
        "def create_tissue_selection(all_genes):\n",
        "    tissue_checkboxes = {}\n",
        "    gene_checkboxes = {}\n",
        "    tissue_accordions = {}\n",
        "\n",
        "    for tissue, genes in all_genes.items():\n",
        "        tissue_checkbox = widgets.Checkbox(\n",
        "            value=False,\n",
        "            description=f\"Select all {tissue} genes\",\n",
        "            indent=False,\n",
        "            layout=widgets.Layout(width='auto')\n",
        "        )\n",
        "        tissue_checkboxes[tissue] = tissue_checkbox\n",
        "\n",
        "        checkboxes = [widgets.Checkbox(description=gene[0]) for gene in genes] #create gene checkboxes\n",
        "        gene_checkboxes.update({f\"{tissue}_{gene[0]}\": checkbox for gene, checkbox in zip(genes, checkboxes)})\n",
        "\n",
        "        grid = widgets.GridBox(\n",
        "            checkboxes,\n",
        "            layout=widgets.Layout(\n",
        "                grid_template_columns='repeat(3, minmax(200px, 1fr))',\n",
        "                grid_gap='10px',\n",
        "                width='100%'\n",
        "            )\n",
        "        )\n",
        "\n",
        "        accordion = widgets.Accordion([grid], titles=[f\"{tissue} Genes\"])\n",
        "        accordion.selected_index = None\n",
        "        tissue_accordions[tissue] = accordion\n",
        "\n",
        "        def on_tissue_change(change, t=tissue):\n",
        "            if change['name'] == 'value':\n",
        "                for gene, checkbox in gene_checkboxes.items():\n",
        "                    if gene.startswith(f\"{t}_\"):\n",
        "                        checkbox.value = change['new']\n",
        "        tissue_checkbox.observe(on_tissue_change)\n",
        "\n",
        "        def on_gene_change(change, t=tissue):\n",
        "            tissue_checkbox.unobserve(on_tissue_change)\n",
        "            tissue_checkbox.value = all(checkbox.value for gene, checkbox in gene_checkboxes.items() if gene.startswith(f\"{t}_\"))\n",
        "            tissue_checkbox.observe(on_tissue_change)\n",
        "\n",
        "        for gene, checkbox in gene_checkboxes.items():\n",
        "            if gene.startswith(f\"{tissue}_\"):\n",
        "                checkbox.observe(lambda change, t=tissue, c=checkbox: on_gene_change(change, t))\n",
        "\n",
        "    tissue_vbox = widgets.VBox([\n",
        "        widgets.HBox([tissue_checkboxes[tissue], tissue_accordions[tissue]])\n",
        "        for tissue in all_genes.keys()\n",
        "    ])\n",
        "\n",
        "    return tissue_vbox, gene_checkboxes\n",
        "\n",
        "def select_all(b):\n",
        "    for checkbox in all_gene_checkboxes.values():\n",
        "        checkbox.value = True\n",
        "\n",
        "def deselect_all(b):\n",
        "    for checkbox in all_gene_checkboxes.values():\n",
        "        checkbox.value = False\n",
        "\n",
        "def get_selected_genes():\n",
        "    return [gene.split('_')[1] for gene, checkbox in all_gene_checkboxes.items() if checkbox.value]\n",
        "\n",
        "def create_gene_list_file(b, taxonomy_dict, taxonomy_id):\n",
        "    \"\"\"Creates gene list file in organized directory structure\"\"\"\n",
        "    selected_genes = get_selected_genes()\n",
        "    if not selected_genes:\n",
        "        print(\"No genes selected. Please select at least one gene.\")\n",
        "        return None\n",
        "\n",
        "    paths = create_project_structure(taxonomy_dict, taxonomy_id)\n",
        "\n",
        "    with open(paths['genelist'], 'w') as f:\n",
        "        for gene in selected_genes:\n",
        "            f.write(f\"{gene}\\n\")\n",
        "\n",
        "    print(f\"Project directory created at: {paths['root']}\")\n",
        "    return paths\n",
        "\n",
        "def create_output_directory(taxonomy_dict, taxonomy_id):\n",
        "    name = taxonomy_dict.get(taxonomy_id, \"Unknown_Taxonomy\")\n",
        "    # Calculate project_name here instead of using the undefined variable\n",
        "    project_name = get_project_name(taxonomy_id, taxonomy_dict, \"Unknown\") # Assuming gene_type is unknown at this point\n",
        "    output_dir = f\"/content/drive/MyDrive/Colab Notebooks/proteoparc/{project_name}\"\n",
        "    os.makedirs(output_dir, exist_ok=True)\n",
        "    return output_dir\n",
        "\n",
        "def get_project_name(taxonomy_id, taxonomy_dict, gene_type):\n",
        "    taxonomy_name = taxonomy_dict.get(taxonomy_id, \"Unknown\")\n",
        "    return f\"{taxonomy_name.lower().replace(' ', '_')}_{gene_type.lower()}_genes\""
      ],
      "metadata": {
        "id": "6sNOKH52j3vl",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title #Select genes and run the analysis\n",
        "\n",
        "# Global variables for UI\n",
        "all_gene_checkboxes = {}\n",
        "\n",
        "# Create UI elements\n",
        "tissue_selection, all_gene_checkboxes = create_tissue_selection(all_genes)\n",
        "select_all_button = widgets.Button(description=\"Select All Genes\")\n",
        "deselect_all_button = widgets.Button(description=\"Deselect All Genes\")\n",
        "print_button = widgets.Button(description=\"Print Selected Genes\")\n",
        "save_button = widgets.Button(description=\"Save Gene List\")\n",
        "run_proteoparc_button = widgets.Button(description=\"Run ProteoParc\")\n",
        "\n",
        "# Attach functions to buttons\n",
        "select_all_button.on_click(select_all)\n",
        "deselect_all_button.on_click(deselect_all)\n",
        "print_button.on_click(lambda b: print(\"Selected genes:\", get_selected_genes()))\n",
        "save_button.on_click(create_gene_list_file)\n",
        "run_proteoparc_button.on_click(run_proteoparc)\n",
        "\n",
        "\n",
        "# Create the button grid\n",
        "button_grid = create_taxonomy_buttons(taxonomy_dict)\n",
        "\n",
        "# Initial display\n",
        "display(HTML(\"<h2>Select Taxonomy</h2>\"))\n",
        "display(button_grid)\n",
        "display(HTML(\"<h2>Gene Selection by Tissue</h2>\"))\n",
        "display(widgets.HBox([select_all_button, deselect_all_button]))\n",
        "display(tissue_selection)\n",
        "display(widgets.HBox([print_button, save_button, run_proteoparc_button]))"
      ],
      "metadata": {
        "id": "2s_KFsocx6J6"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}