{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "private_outputs": true,
      "provenance": [],
      "mount_file_id": "1TgC0Cd6R7STlydfr1Vg95vDVFpiUBnFI",
      "authorship_tag": "ABX9TyOqbw3+TGmyWf+ZeW8JKJBv",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Palaeoprot/ProteoParc_Colab/blob/main/ProteoParc_in_Colab.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##License\n",
        "\n",
        "The [ProteoParc](https://github.com/guillecarrillo/proteoparc) pipeline was written by [Guillermo Carrillo Mart√≠n](https://github.com/guillecarrillo) and this is his license from Github. \"[Palaeoparc] \"*is the result of many collective work, so its citation in any study in which it is used would be appreciated.\n",
        "\n",
        "\n",
        " Every user is free to change any section of the code so it can be adjusted to individual necessities, if you do so, please indicate it in your manuscripts writing the changes to avoid confusion. For more information, read the License This pi[link text](https://)peline has been possible thanks to a lot of collective work, so its citation in any study in which it is used would be appreciated. Every user is free to change any section of the code so it can be adjusted to individual necessities, if you do so, please indicate it in your manuscripts to avoid confusion.*\"\n",
        "\n",
        "Here is my take on it in Colab\n",
        "\n",
        "<img src='https://drive.google.com/uc?export=view&id=1FqSr3E9UQKtRO4KwR6LjAU8TPrD8Oq6q' width=800px align=centre>\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "73G4TMMXt0TO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython\n",
        "import requests, sys, json, os, time  #data handling\n",
        "import argparse  # command-line arguments\n",
        "import shlex  # parsing shell commands\n",
        "import subprocess  # For system commands (already imported twice)\n",
        "import ipywidgets as widgets  # For interactive elements\n",
        "from IPython.display import display, HTML, clear_output  # For interactive output\n",
        "import shutil\n",
        "import re\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from requests.adapters import HTTPAdapter, Retry\n",
        "from datetime import datetime\n",
        "# Mount Google Drive (Optional, you can comment out if not using)\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "# Global constants\n",
        "WEBSITE_API = \"https://rest.uniprot.org/\""
      ],
      "metadata": {
        "id": "xJAaQwqisbMX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ##Load the dictionaries\n",
        "# List of gene names for different tissues\n",
        "bone_protein_genes = [\n",
        "    (\"ALB\", \"serum albumin\"),\n",
        "    (\"AEBP1\", \"AE binding protein 1\"),\n",
        "    (\"AHSG\", \"alpha-2-HS-glycoprotein (fetuin-A)\"),\n",
        "    (\"ALPL\", \"alkaline phosphatase, liver/bone/kidney\"),\n",
        "    (\"APOA1\", \"apolipoprotein A-I\"),\n",
        "    (\"APOA4\", \"apolipoprotein A-IV\"),\n",
        "    (\"APOE\", \"apolipoprotein E\"),\n",
        "    (\"APOC1\", \"apolipoprotein C-I\"),\n",
        "    (\"APP\", \"amyloid beta precursor protein\"),\n",
        "    (\"ASPN\", \"asporin\"),\n",
        "    (\"BGLAP\", \"bone gamma-carboxyglutamate protein (osteocalcin)\"),\n",
        "    (\"BGN\", \"biglycan\"),\n",
        "    (\"C3\", \"complement component 3\"),\n",
        "    (\"C8B\", \"complement component 8, beta polypeptide\"),\n",
        "    (\"C9\", \"complement component 9\"),\n",
        "    (\"CFH\", \"complement factor H\"),\n",
        "    (\"CHAD\", \"chondroadherin\"),\n",
        "    (\"CHGA\", \"chromogranin A\"),\n",
        "    (\"CLEC11A\", \"C-type lectin domain family 11, member A\"),\n",
        "    (\"CLEC3B\", \"C-type lectin domain family 3, member B\"),\n",
        "    (\"CLU\", \"clusterin\"),\n",
        "    (\"COL10A1\", \"collagen type X alpha 1 chain\"),\n",
        "    (\"COL11A1\", \"collagen type XI alpha 1 chain\"),\n",
        "    (\"COL11A2\", \"collagen type XI alpha 2 chain\"),\n",
        "    (\"COL12A1\", \"collagen type XII alpha 1 chain\"),\n",
        "    (\"COL16A1\", \"collagen type XVI alpha 1 chain\"),\n",
        "    (\"COL1A1\", \"collagen type I alpha 1 chain\"),\n",
        "    (\"COL1A2\", \"collagen type I alpha 2 chain\"),\n",
        "    (\"COL21A1\", \"collagen type XXI alpha 1 chain\"),\n",
        "    (\"COL22A1\", \"collagen type XXII alpha 1 chain\"),\n",
        "    (\"COL2A1\", \"collagen type II alpha 1 chain\"),\n",
        "    (\"COL3A1\", \"collagen type III alpha 1 chain\"),\n",
        "    (\"COL4A3\", \"collagen type IV alpha 3 chain\"),\n",
        "    (\"COL4A4\", \"collagen type IV alpha 4 chain\"),\n",
        "    (\"COL4A5\", \"collagen type IV alpha 5 chain\"),\n",
        "    (\"COL5A1\", \"collagen type V alpha 1 chain\"),\n",
        "    (\"COL5A2\", \"collagen type V alpha 2 chain\"),\n",
        "    (\"COL5A3\", \"collagen type V alpha 3 chain\"),\n",
        "    (\"COL6A1\", \"collagen type VI alpha 1 chain\"),\n",
        "    (\"COL6A3\", \"collagen type VI alpha 3 chain\"),\n",
        "    (\"COL8A1\", \"collagen type VIII alpha 1 chain\"),\n",
        "    (\"CRP\", \"C-reactive protein\"),\n",
        "    (\"DCN\", \"decorin\"),\n",
        "    (\"DPT\", \"dermatopontin\"),\n",
        "    (\"EEF1A1\", \"eukaryotic translation elongation factor 1 alpha 1\"),\n",
        "    (\"EMILIN1\", \"elastin microfibril interfacer 1\"),\n",
        "    (\"EZR\", \"ezrin\"),\n",
        "    (\"F10\", \"coagulation factor X\"),\n",
        "    (\"F2\", \"coagulation factor II (thrombin)\"),\n",
        "    (\"F7\", \"coagulation factor VII\"),\n",
        "    (\"F9\", \"coagulation factor IX\"),\n",
        "    (\"FGL2\", \"fibrinogen-like 2\"),\n",
        "    (\"FMOD\", \"fibromodulin\"),\n",
        "    (\"FN1\", \"fibronectin 1\"),\n",
        "    (\"GAPDH\", \"glyceraldehyde-3-phosphate dehydrogenase\"),\n",
        "    (\"GAS6\", \"growth arrest-specific 6\"),\n",
        "    (\"GC\", \"group-specific component (vitamin D binding protein)\"),\n",
        "    (\"HAPLN3\", \"hyaluronan and proteoglycan link protein 3\"),\n",
        "    (\"HSP90B1\", \"heat shock protein 90 beta family member 1\"),\n",
        "    (\"HSPA5\", \"heat shock protein family A (Hsp70) member 5\"),\n",
        "    (\"HTRA1\", \"HtrA serine peptidase 1\"),\n",
        "    (\"IBSP\", \"integrin binding sialoprotein\"),\n",
        "    (\"IGF1\", \"insulin-like growth factor 1\"),\n",
        "    (\"IGF2\", \"insulin-like growth factor 2\"),\n",
        "    (\"IGFALS\", \"insulin-like growth factor binding protein, acid labile subunit\"),\n",
        "    (\"IGFBP1\", \"insulin-like growth factor binding protein 1\"),\n",
        "    (\"KNG1\", \"kininogen 1\"),\n",
        "    (\"KRT2\", \"keratin 2\"),\n",
        "    (\"LOX\", \"lysyl oxidase\"),\n",
        "    (\"LRRC15\", \"leucine rich repeat containing 15\"),\n",
        "    (\"LUM\", \"lumican\"),\n",
        "    (\"MGP\", \"matrix Gla protein\"),\n",
        "    (\"MMP2\", \"matrix metallopeptidase 2\"),\n",
        "    (\"MSN\", \"moesin\"),\n",
        "    (\"MYO1B\", \"myosin IB\"),\n",
        "    (\"NUCB1\", \"nucleobindin 1\"),\n",
        "    (\"NUCB2\", \"nucleobindin 2\"),\n",
        "    (\"OGN\", \"osteoglycin\"),\n",
        "    (\"OLFML1\", \"olfactomedin-like 1\"),\n",
        "    (\"OLFML3\", \"olfactomedin-like 3\"),\n",
        "    (\"OMD\", \"osteomodulin\"),\n",
        "    (\"P4HB\", \"prolyl 4-hydroxylase subunit beta\"),\n",
        "    (\"PAM\", \"peptidylglycine alpha-amidating monooxygenase\"),\n",
        "    (\"PCOLCE\", \"procollagen C-endopeptidase enhancer\"),\n",
        "    (\"PHOSPHO1\", \"phosphatase, orphan 1\"),\n",
        "    (\"POSTN\", \"periostin\"),\n",
        "    (\"PROC\", \"protein C, inactivator of coagulation factors Va and VIIIa\"),\n",
        "    (\"PROS1\", \"protein S, vitamin K-dependent\"),\n",
        "    (\"PRSS2\", \"protease, serine 2\"),\n",
        "    (\"SERPINC1\", \"serpin family C member 1 (antithrombin)\"),\n",
        "    (\"SERPIND1\", \"serpin family D member 1 (heparin cofactor II)\"),\n",
        "    (\"SERPINF1\", \"serpin family F member 1\"),\n",
        "    (\"SLC8A3\", \"solute carrier family 8 member A3\"),\n",
        "    (\"SPARC\", \"secreted protein acidic and rich in cysteine (osteonectin)\"),\n",
        "    (\"SPARCL1\", \"SPARC-like 1 (hevin)\"),\n",
        "    (\"SPP1\", \"secreted phosphoprotein 1 (osteopontin)\"),\n",
        "    (\"SPP2\", \"secreted phosphoprotein 2\"),\n",
        "    (\"TGFB1\", \"transforming growth factor beta 1\"),\n",
        "    (\"THBS1\", \"thrombospondin 1\"),\n",
        "    (\"TNC\", \"tenascin C\"),\n",
        "    (\"TPP1\", \"tripeptidyl peptidase I\"),\n",
        "    (\"TUBA1B\", \"tubulin alpha 1b\"),\n",
        "    (\"VCAN\", \"versican\"),\n",
        "    (\"VIT\", \"vitrin\"),\n",
        "    (\"VTN\", \"vitronectin\")\n",
        "]\n",
        "\n",
        "enamel_protein_genes = [\n",
        "    (\"ALB\", \"serum albumin\"),\n",
        "    (\"COL1A2\", \"collagen type I alpha 2 chain\"),\n",
        "    (\"COL1A1\", \"collagen type I alpha 1 chain\"),\n",
        "    (\"COL2A1\", \"collagen type II alpha 1 chain\"),\n",
        "    (\"AMBN\", \"ameloblastin\"),\n",
        "    (\"AMELX\", \"amelogenin X-linked\"),\n",
        "    (\"ENAM\", \"enamelin\"),\n",
        "    (\"AMELY\", \"amelogenin Y-linked\"),\n",
        "    (\"AMTN\", \"amelotin\"),\n",
        "    (\"MMP20\", \"matrix metallopeptidase 20\"),\n",
        "    (\"KLK4\", \"kallikrein related peptidase 4\"),\n",
        "    (\"TUFT1\", \"tuftelin 1\"),\n",
        "    (\"SERPINC1\", \"serpin family C member 1 (antithrombin)\"),\n",
        "    (\"ODAM\", \"odontogenic, ameloblast associated\"),\n",
        "    (\"COL17A1\", \"collagen type XVII alpha 1 chain\")\n",
        "]\n",
        "\n",
        "food_crust_protein_genes = [\n",
        "    (\"CSN1S1\", \"alpha-S1 casein\"),\n",
        "    (\"CSN1S2\", \"alpha-S2 casein\"),\n",
        "    (\"CSN2\", \"beta-casein\"),\n",
        "    (\"CSN3\", \"kappa-casein\"),\n",
        "    (\"LALBA\", \"alpha-lactalbumin\"),\n",
        "    (\"LGB\", \"beta-lactoglobulin\"),\n",
        "    (\"IGH@\", \"immunoglobulin heavy chains\"),\n",
        "    (\"IGK@\", \"immunoglobulin kappa light chains\"),\n",
        "    (\"IGL@\", \"immunoglobulin lambda light chains\"),\n",
        "    (\"LTF\", \"lactoferrin\"),\n",
        "    (\"LYZ\", \"lysozyme\"),\n",
        "    (\"ALB\", \"serum albumin\"),\n",
        "    (\"MUC1\", \"mucin 1\"),\n",
        "    (\"BTN1A1\", \"butyrophilin subfamily 1 member A1\"),\n",
        "    (\"XDH\", \"xanthine dehydrogenase/oxidase\"),\n",
        "    (\"MFGE8\", \"milk fat globule-EGF factor 8 protein\")\n",
        "]\n",
        "\n",
        "zooms_genes = [\n",
        "    (\"COL1A2\", \"collagen type I alpha 2 chain\"),\n",
        "    (\"COL1A1\", \"collagen type I alpha 1 chain\"),\n",
        "    (\"COL3A1\", \"collagen type III alpha 1 chain\")\n",
        "]\n",
        "\n",
        "# Define a combined dictionary of all genes\n",
        "all_genes = {\n",
        "    \"Bone\": bone_protein_genes,\n",
        "    \"Enamel\": enamel_protein_genes,\n",
        "    \"Food Crust\": food_crust_protein_genes,\n",
        "    \"Zooms\": zooms_genes\n",
        "}\n",
        "\n",
        "# Define the taxonomy IDs and their names\n",
        "taxonomy_dict = {\n",
        "    # Carnivora\n",
        "    33554: \"Carnivora\", 379584: \"Caniformia\", 379583: \"Feliformia\", 9632: \"Ursidae\", 9615: \"Canis lupus familiaris\",\n",
        "\n",
        "    # Lagomorpha\n",
        "    9975: \"Lagomorpha\",\n",
        "\n",
        "    # Humans\n",
        "    9606: \"Homo sapiens\", 9605: \"Homo\", 9604: \"Hominidae\", 9443: \"Primates\", 40674: \"Mammalia\", 7711: \"Chordata\", 33208: \"Animalia\",\n",
        "\n",
        "    # Ostriches\n",
        "    8839: \"Struthio camelus\", 8838: \"Struthio\", 8837: \"Struthionidae\", 30498: \"Struthioniformes\", 8782: \"Aves\",\n",
        "\n",
        "    # Cod\n",
        "    8049: \"Gadus morhua\", 8048: \"Gadus\", 27502: \"Gadidae\", 27682: \"Gadiformes\", 7898: \"Actinopterygii\",\n",
        "\n",
        "    # Geckos\n",
        "    103697: \"Gekko gecko\", 8489: \"Gekko\", 8483: \"Gekkonidae\", 8509: \"Squamata\", 8457: \"Reptilia\"\n",
        "}\n",
        "\n",
        "# print(\"all_genes:\")\n",
        "# for tissue, genes in all_genes.items():\n",
        "#     print(f\"  {tissue}: {len(genes)} genes\")\n",
        "\n",
        "# print(\"\\ntaxonomy_dict:\")\n",
        "# for tax_id, tax_name in taxonomy_dict.items():\n",
        "#     print(f\"  {tax_id}: {tax_name}\")"
      ],
      "metadata": {
        "cellView": "form",
        "id": "v-hYV_bKsdqm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ##Functions\n",
        "import os\n",
        "import re\n",
        "import shutil\n",
        "import pandas as pd\n",
        "import requests\n",
        "from requests.adapters import HTTPAdapter, Retry\n",
        "from datetime import datetime\n",
        "from Bio import SeqIO\n",
        "import glob\n",
        "\n",
        "# Global variables\n",
        "WEBSITE_API = \"https://rest.uniprot.org/\"\n",
        "\n",
        "def setup_api_session():\n",
        "    retries = Retry(total=5, backoff_factor=0.25, status_forcelist=[500, 502, 503, 504])\n",
        "    session = requests.Session()\n",
        "    session.mount(\"https://\", HTTPAdapter(max_retries=retries))\n",
        "    return session\n",
        "\n",
        "def get_next_link(headers):\n",
        "    re_next_link = re.compile(r'<(.+)>; rel=\"next\"')\n",
        "    if \"Link\" in headers:\n",
        "        match = re_next_link.match(headers[\"Link\"])\n",
        "        if match:\n",
        "            return match.group(1)\n",
        "\n",
        "def get_batch(session, batch_url):\n",
        "    while batch_url:\n",
        "        response = session.get(batch_url)\n",
        "        response.raise_for_status()\n",
        "        total = response.headers[\"x-total-results\"]\n",
        "        yield response, total\n",
        "        batch_url = get_next_link(response.headers)\n",
        "\n",
        "def json_to_fasta(record, tax_id_list, gene_name=None):\n",
        "    for database in record[\"uniParcCrossReferences\"]:\n",
        "        if \"organism\" in database and database[\"organism\"][\"taxonId\"] in tax_id_list:\n",
        "            header = f\">{database['database']}|{record['uniParcId']}|{database['lastUpdated']} \"\n",
        "            if \"proteinName\" in database:\n",
        "                header += f\"{database['proteinName']} \"\n",
        "            header += f\"OS={database['organism']['scientificName']} OX={database['organism']['taxonId']}\"\n",
        "            if gene_name:\n",
        "                header += f\" GN={gene_name}\"\n",
        "            elif \"geneName\" in database:\n",
        "                header += f\" GN={database['geneName']}\"\n",
        "            header += f\" SV={record['uniParcCrossReferences'][0]['versionI']}\"\n",
        "            sequence = record[\"sequence\"][\"value\"]\n",
        "            return header.replace(',', '').replace(';', '').replace(':', ''), sequence\n",
        "    return None, None\n",
        "\n",
        "def download_proteins(tax_id, gene_list, output_file):\n",
        "    session = setup_api_session()\n",
        "    tax_id_list = get_taxid_descendants(tax_id)\n",
        "\n",
        "    with open(output_file, \"w\") as file:\n",
        "        if gene_list:\n",
        "            for gene in gene_list:\n",
        "                url = f\"{WEBSITE_API}uniparc/search?compressed=false&format=json&query=((gene:{gene}) AND (taxonomy_id:{tax_id}))&size=500\"\n",
        "                process_batch(session, url, file, tax_id_list, gene)\n",
        "        else:\n",
        "            url = f\"{WEBSITE_API}uniparc/search?compressed=false&format=json&query=((taxonomy_id:{tax_id}))&size=500\"\n",
        "            process_batch(session, url, file, tax_id_list)\n",
        "\n",
        "def process_batch(session, url, file, tax_id_list, gene=None):\n",
        "    for batch, total in get_batch(session, url):\n",
        "        json_retrieve = batch.json()\n",
        "        for protein in json_retrieve[\"results\"]:\n",
        "            header, sequence = json_to_fasta(protein, tax_id_list, gene)\n",
        "            if header and sequence:\n",
        "                file.write(f\"{header}\\n\")\n",
        "                for i in range(0, len(sequence), 60):\n",
        "                    file.write(f\"{sequence[i:i+60]}\\n\")\n",
        "\n",
        "def get_taxid_descendants(tax_id):\n",
        "    url = f\"{WEBSITE_API}taxonomy/stream?format=list&query=((ancestor:{tax_id}))\"\n",
        "    response = requests.get(url)\n",
        "    tax_id_list = [int(tid) for tid in response.text.strip().split(\"\\n\") if tid]\n",
        "    return tax_id_list\n",
        "\n",
        "def remove_duplicates(input_file, output_folder):\n",
        "    records = []\n",
        "    with open(input_file, \"r\") as f:\n",
        "        for record in SeqIO.parse(f, \"fasta\"):\n",
        "            records.append(record)\n",
        "\n",
        "    unique_records = []\n",
        "    duplicate_records = []\n",
        "    sequences = set()\n",
        "\n",
        "    for record in records:\n",
        "        if str(record.seq) not in sequences:\n",
        "            sequences.add(str(record.seq))\n",
        "            unique_records.append(record)\n",
        "        else:\n",
        "            duplicate_records.append(record)\n",
        "\n",
        "    filtered_file = os.path.join(output_folder, \"filtered_mf.fasta\")\n",
        "    duplicates_file = os.path.join(output_folder, \"duplicate_records.fasta\")\n",
        "\n",
        "    SeqIO.write(unique_records, filtered_file, \"fasta\")\n",
        "    SeqIO.write(duplicate_records, duplicates_file, \"fasta\")\n",
        "\n",
        "    return len(duplicate_records)\n",
        "\n",
        "import os\n",
        "import re\n",
        "import shutil\n",
        "import pandas as pd\n",
        "import requests\n",
        "from requests.adapters import HTTPAdapter, Retry\n",
        "from datetime import datetime\n",
        "from Bio import SeqIO\n",
        "import glob\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display, HTML, clear_output\n",
        "\n",
        "# Global variables\n",
        "WEBSITE_API = \"https://rest.uniprot.org/\"\n",
        "\n",
        "def create_taxonomy_buttons(taxonomy_dict):\n",
        "    buttons = []\n",
        "    for tax_id, name in taxonomy_dict.items():\n",
        "        button = widgets.Button(\n",
        "            description=name,\n",
        "            tooltip=f\"{name} (ID: {tax_id})\",\n",
        "            layout=widgets.Layout(width='auto', height='auto'),\n",
        "            style={'button_color': '#e7e7e7', 'font_weight': 'normal'}\n",
        "        )\n",
        "\n",
        "        def on_button_click(b, tax_id=tax_id, name=name):\n",
        "            global selected_taxonomy_id\n",
        "            selected_taxonomy_id = tax_id\n",
        "            output_dir = create_output_directory(taxonomy_dict, tax_id)\n",
        "            create_gene_list_file.output_dir = output_dir  # Store output_dir for later use\n",
        "            run_proteoparc.taxonomy_id = tax_id  # Store taxonomy_id for later use\n",
        "            url = f\"https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id={tax_id}\"\n",
        "            clear_output()\n",
        "            display(HTML(\"<h2>Selected Taxonomy</h2>\"))\n",
        "            display(HTML(f\"<h3>{name} (ID: {tax_id})</h3>\"))\n",
        "            display(HTML(f\"<a href='{url}' target='_blank'>Open NCBI Taxonomy Browser</a>\"))\n",
        "            display(HTML(f\"<p>Output directory: {output_dir}</p>\"))\n",
        "            display(HTML(\"<h2>Select Taxonomy</h2>\"))\n",
        "            display(button_grid)\n",
        "            display(HTML(\"<h2>Gene Selection by Tissue</h2>\"))\n",
        "            display(widgets.HBox([select_all_button, deselect_all_button]))\n",
        "            display(tissue_selection)\n",
        "            display(widgets.HBox([print_button, save_button, run_proteoparc_button]))\n",
        "\n",
        "        button.on_click(on_button_click)\n",
        "        buttons.append(button)\n",
        "\n",
        "    button_grid = widgets.GridBox(\n",
        "        buttons,\n",
        "        layout=widgets.Layout(\n",
        "            grid_template_columns='repeat(3, auto)',\n",
        "            grid_gap='10px',\n",
        "            width='100%'\n",
        "        )\n",
        "    )\n",
        "\n",
        "    return button_grid\n",
        "\n",
        "def create_gene_checkboxes(gene_list):\n",
        "    return [widgets.Checkbox(\n",
        "        value=False,\n",
        "        description=f\"{gene} ({desc})\",\n",
        "        indent=False,\n",
        "        layout=widgets.Layout(width='auto')\n",
        "    ) for gene, desc in gene_list]\n",
        "\n",
        "def create_tissue_selection(all_genes):\n",
        "    tissue_checkboxes = {}\n",
        "    gene_checkboxes = {}\n",
        "    tissue_accordions = {}\n",
        "\n",
        "    for tissue, genes in all_genes.items():\n",
        "        tissue_checkbox = widgets.Checkbox(\n",
        "            value=False,\n",
        "            description=f\"Select all {tissue} genes\",\n",
        "            indent=False,\n",
        "            layout=widgets.Layout(width='auto')\n",
        "        )\n",
        "        tissue_checkboxes[tissue] = tissue_checkbox\n",
        "\n",
        "        checkboxes = create_gene_checkboxes(genes)\n",
        "        gene_checkboxes.update({f\"{tissue}_{gene}\": checkbox for (gene, _), checkbox in zip(genes, checkboxes)})\n",
        "\n",
        "        grid = widgets.GridBox(\n",
        "            checkboxes,\n",
        "            layout=widgets.Layout(\n",
        "                grid_template_columns='repeat(3, minmax(200px, 1fr))',\n",
        "                grid_gap='10px',\n",
        "                width='100%'\n",
        "            )\n",
        "        )\n",
        "\n",
        "        accordion = widgets.Accordion([grid], titles=[f\"{tissue} Genes\"])\n",
        "        accordion.selected_index = None\n",
        "        tissue_accordions[tissue] = accordion\n",
        "\n",
        "        def on_tissue_change(change, t=tissue):\n",
        "            if change['name'] == 'value':\n",
        "                for gene, checkbox in gene_checkboxes.items():\n",
        "                    if gene.startswith(f\"{t}_\"):\n",
        "                        checkbox.value = change['new']\n",
        "        tissue_checkbox.observe(on_tissue_change)\n",
        "\n",
        "        def on_gene_change(change, t=tissue):\n",
        "            tissue_checkbox.unobserve(on_tissue_change)\n",
        "            tissue_checkbox.value = all(checkbox.value for gene, checkbox in gene_checkboxes.items() if gene.startswith(f\"{t}_\"))\n",
        "            tissue_checkbox.observe(on_tissue_change)\n",
        "        for gene, checkbox in gene_checkboxes.items():\n",
        "            if gene.startswith(f\"{tissue}_\"):\n",
        "                checkbox.observe(on_gene_change)\n",
        "\n",
        "    tissue_vbox = widgets.VBox([\n",
        "        widgets.HBox([tissue_checkboxes[tissue], tissue_accordions[tissue]])\n",
        "        for tissue in all_genes.keys()\n",
        "    ])\n",
        "\n",
        "    return tissue_vbox, gene_checkboxes\n",
        "\n",
        "def select_all(b):\n",
        "    for checkbox in all_gene_checkboxes.values():\n",
        "        checkbox.value = True\n",
        "\n",
        "def deselect_all(b):\n",
        "    for checkbox in all_gene_checkboxes.values():\n",
        "        checkbox.value = False\n",
        "\n",
        "def get_selected_genes():\n",
        "    return [gene.split('_')[1] for gene, checkbox in all_gene_checkboxes.items() if checkbox.value]\n",
        "\n",
        "def create_gene_list_file(b, custom_path=None):\n",
        "    selected_genes = get_selected_genes()\n",
        "    if not selected_genes:\n",
        "        print(\"No genes selected. Please select at least one gene.\")\n",
        "        return None\n",
        "\n",
        "    file_path = custom_path or os.path.join(create_gene_list_file.output_dir, \"genelist.txt\")\n",
        "    with open(file_path, 'w') as f:\n",
        "        for gene in selected_genes:\n",
        "            f.write(f\"{gene}\\n\")\n",
        "    print(f\"Gene list has been saved to {file_path}\")\n",
        "    return file_path\n",
        "\n",
        "def create_output_directory(taxonomy_dict, taxonomy_id):\n",
        "    name = taxonomy_dict.get(taxonomy_id, \"Unknown_Taxonomy\")\n",
        "    # Calculate project_name here instead of using the undefined variable\n",
        "    project_name = get_project_name(taxonomy_id, taxonomy_dict, \"Unknown\") # Assuming gene_type is unknown at this point\n",
        "    output_dir = f\"/content/drive/MyDrive/Colab Notebooks/proteoparc/{project_name}\"\n",
        "    os.makedirs(output_dir, exist_ok=True)\n",
        "    return output_dir\n",
        "\n",
        "def get_project_name(taxonomy_id, taxonomy_dict, gene_type):\n",
        "    taxonomy_name = taxonomy_dict.get(taxonomy_id, \"Unknown\")\n",
        "    return f\"{taxonomy_name.lower().replace(' ', '_')}_{gene_type.lower()}_genes\"\n",
        "\n",
        "def run_proteoparc(b):\n",
        "    if not hasattr(create_gene_list_file, 'output_dir') or not hasattr(run_proteoparc, 'taxonomy_id'):\n",
        "        print(\"Please select a taxonomy first.\")\n",
        "        return\n",
        "\n",
        "    output_dir = create_gene_list_file.output_dir\n",
        "\n",
        "    gene_list_path = os.path.join(output_dir, \"genelist.txt\")\n",
        "    create_gene_list_file(None, gene_list_path)\n",
        "\n",
        "    if not os.path.exists(gene_list_path):\n",
        "        print(f\"Error: Gene list file not created at {gene_list_path}\")\n",
        "        return\n",
        "\n",
        "    print(\"Gene list contents:\", get_selected_genes())\n",
        "\n",
        "    selected_genes = get_selected_genes()\n",
        "    gene_type = \"ZooMS\" if any(gene in selected_genes for gene in all_genes.get(\"ZooMS\", [])) else \"mixed\"\n",
        "\n",
        "    project_name = get_project_name(run_proteoparc.taxonomy_id, taxonomy_dict, gene_type)\n",
        "    output_dir = f\"/content/proteoparc/{project_name}\"\n",
        "    os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "    print(\"Running ProteoParc pipeline...\")\n",
        "\n",
        "    run_integrated_proteoparc(run_proteoparc.taxonomy_id, gene_list_path, output_dir)\n",
        "\n",
        "    if os.path.exists(output_dir):\n",
        "        print(f\"Output directory: {output_dir}\")\n",
        "        fasta_files = glob.glob(os.path.join(output_dir, \"**\", \"*.fasta\"), recursive=True)\n",
        "        if fasta_files:\n",
        "            print(f\"Found {len(fasta_files)} FASTA file(s):\")\n",
        "            for fasta_file in fasta_files:\n",
        "                print(f\"  - {os.path.basename(fasta_file)}\")\n",
        "        else:\n",
        "            print(\"No FASTA files found in the output directory.\")\n",
        "    else:\n",
        "        print(f\"Output directory not found: {output_dir}\")\n",
        "\n",
        "    print(\"\\nSummary of gene search results:\")\n",
        "    gene_list = get_selected_genes()\n",
        "    for gene in gene_list:\n",
        "        found = any(gene in open(f).read() for f in fasta_files)\n",
        "        print(f\"  - {gene}: {'Found' if found else 'Not found'}\")\n",
        "\n",
        "    colab_output_dir = f\"/content/drive/MyDrive/Colab Notebooks/proteoparc/{project_name}\"\n",
        "    if os.path.exists(output_dir):\n",
        "        shutil.copytree(output_dir, colab_output_dir, dirs_exist_ok=True)\n",
        "        print(f\"\\nResults copied to Colab environment: {colab_output_dir}\")\n",
        "    else:\n",
        "        print(\"\\nNo results to copy to Colab environment\")\n",
        "\n",
        "def run_integrated_proteoparc(tax_id, gene_list_path, output_folder):\n",
        "    os.makedirs(output_folder, exist_ok=True)\n",
        "\n",
        "    # Step 1: Download proteins\n",
        "    raw_fasta = os.path.join(output_folder, \"raw_proteins.fasta\")\n",
        "    gene_list = []\n",
        "    if gene_list_path:\n",
        "        with open(gene_list_path, \"r\") as f:\n",
        "            gene_list = [line.strip() for line in f]\n",
        "    download_proteins(tax_id, gene_list, raw_fasta)\n",
        "\n",
        "    # Step 2: Remove duplicates\n",
        "    duplicates_removed = remove_duplicates(raw_fasta, output_folder)\n",
        "    filtered_fasta = os.path.join(output_folder, \"filtered_mf.fasta\")\n",
        "\n",
        "    # Step 3: Generate metadata\n",
        "    metadata_folder = os.path.join(output_folder, \"metadata\")\n",
        "    generate_metadata(filtered_fasta, metadata_folder, gene_list_path)\n",
        "\n",
        "    print(f\"ProteoParc pipeline completed. Output folder: {output_folder}\")\n",
        "    print(f\"Duplicates removed: {duplicates_removed}\")\n"
      ],
      "metadata": {
        "id": "Ms-HV2jyxzmZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ##Select genes and run the analysis\n",
        "\n",
        "# Global variables for UI\n",
        "all_gene_checkboxes = {}\n",
        "\n",
        "# Create UI elements\n",
        "tissue_selection, all_gene_checkboxes = create_tissue_selection(all_genes)\n",
        "select_all_button = widgets.Button(description=\"Select All Genes\")\n",
        "deselect_all_button = widgets.Button(description=\"Deselect All Genes\")\n",
        "print_button = widgets.Button(description=\"Print Selected Genes\")\n",
        "save_button = widgets.Button(description=\"Save Gene List\")\n",
        "run_proteoparc_button = widgets.Button(description=\"Run ProteoParc\")\n",
        "\n",
        "# Attach functions to buttons\n",
        "select_all_button.on_click(select_all)\n",
        "deselect_all_button.on_click(deselect_all)\n",
        "print_button.on_click(lambda b: print(\"Selected genes:\", get_selected_genes()))\n",
        "save_button.on_click(create_gene_list_file)\n",
        "run_proteoparc_button.on_click(run_proteoparc)\n",
        "\n",
        "\n",
        "# Create the button grid\n",
        "button_grid = create_taxonomy_buttons(taxonomy_dict)\n",
        "\n",
        "# Initial display\n",
        "display(HTML(\"<h2>Select Taxonomy</h2>\"))\n",
        "display(button_grid)\n",
        "display(HTML(\"<h2>Gene Selection by Tissue</h2>\"))\n",
        "display(widgets.HBox([select_all_button, deselect_all_button]))\n",
        "display(tissue_selection)\n",
        "display(widgets.HBox([print_button, save_button, run_proteoparc_button]))"
      ],
      "metadata": {
        "id": "2s_KFsocx6J6"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}